src/Regulator.cpp

```diff
@@ -648,41 +649,71 @@ bool Regulator::pullPacket()
                     continue;
                 updatePushStats(next);
                 // count how many we have skipped
-                skipped = next - mLastSeqNumOut - 1;
-                if (skipped < 0)
-                    skipped += NumSlots;
+                numSkipped = next - mLastSeqNumOut - 1;
+                if (numSkipped < 0)
+                    numSkipped += NumSlots;
             }
             // check if packet's age matches tolerance, or is the best candidate we have
-            if (mIncomingTiming[next] + mMsecTolerance >= now || i == 0) {
+            const bool meetsTolerance = mIncomingTiming[next] + mMsecTolerance >= now;
+            if (meetsTolerance || i == 0) {
                 // next is the best candidate
-                memcpy(mXfrBuffer, mSlots[next], mPeerBytes);
-                mLastSeqNumOut = next;
-                double latency = (now - mIncomingTiming[mLastSeqNumOut]);
-                if (latency > mStatsMaxLatency) {
-                    mStatsMaxLatency = latency;
+                if (numSkipped > 0) {
+                    // if we skipped any packets, process it as a glitch
+                    if (meetsTolerance) {
+                        // potentially, we can use next on the next pass
+                        if (prevSkipped != -1) {
+                            // increment last out to the previous valid skipped packet
+                            // which would have been used if it met tolerance
+                            numSkipped = prevSkipped - mLastSeqNumOut - 1;
+                            if (numSkipped < 0)
+                                numSkipped += NumSlots;
+                            mLastSeqNumOut = prevSkipped;
+                            // use the previous valid skipped packet for training
+                            memcpy(mXfrBuffer, mSlots[mLastSeqNumOut], mPeerBytes);
+                        } else {
+                            // increment last out to the previous packet
+                            numSkipped--;
+                            mLastSeqNumOut = next - 1;
+                            if (mLastSeqNumOut < 0)
+                                mLastSeqNumOut = NumSlots;
+                            // use the next "real" packet for training
+                            memcpy(mXfrBuffer, mSlots[next], mPeerBytes);
+                        }
+                    } else {
+                        // next packet doesn't meet tolerance
+                        mLastSeqNumOut = next;
+                        memcpy(mXfrBuffer, mSlots[mLastSeqNumOut], mPeerBytes);
+                    }
+                    pullStat->plcOverruns += numSkipped;
+                    pullStat->plcUnderruns++;
+                    processPacket(true);
+                    return true;
+                } else {
+                    // none skipped: next is a "real" packet (not a glitch)
```

Comment:
Note that if no slots are being skipped, this block is called. It's the same logic we currently have with the "PACKETOK" and "OUTPUT" gotos. I'm just eliminating those because they are not used elsewhere, and I think gotos in general are hard to read and should generally be avoided.

Reply:
note that it's best to use a fixed queue size when testing (i.e. `-q5`) that includes enough headroom to get "extra fresh" packets into the mix.